server/branch.ts

```diff
@@ -0,0 +1,257 @@
+	"Noop": NoopBranch;
+}
+type QuestionBranch = QuestionBranchTypes[keyof QuestionBranchTypes];
+
+const SCHEMA_FILE: string = "./config/questions.schema.json";
+
+export class NoopBranch {
+	public readonly name: string;
+	public readonly type: keyof QuestionBranchTypes = "Noop";
+
+	public textBlocks: TextBlocks | undefined;
+	public questions: Questions;
+	public questionLabels: Labels;
+
+	protected readonly location: string;
+
+	constructor(name: string, location: string) {
+		this.name = name;
+		this.location = location;
+	}
+
+	protected async loadSettings(): Promise<void> {
+		return;
+	}
+	protected serializeSettings(): QuestionBranchSettings {
+		// Noop branch has no settings
+		return {};
+	}
+
+	public async loadFromSchema(): Promise<this> {
+		let questionBranches: QuestionBranches = JSON.parse(await readFileAsync(this.location));
+		let schema = JSON.parse(await readFileAsync(path.resolve(__dirname, SCHEMA_FILE)));
+
+		let validator = new ajv();
+		let valid = validator.validate(schema, questionBranches);
+		let branchNames = questionBranches.map(branch => branch.name);
+		let questionBranch = questionBranches.find(branch => branch.name === this.name);
+		if (!valid) {
+			throw new Error(JSON.stringify(validator.errors));
+		}
+		else if (new Set(branchNames).size !== branchNames.length) {
+			throw new Error("Application branch names are not unique");
+		}
+		else if (!questionBranch) {
+			throw new Error(`Branch "${this.name}" not found in schema (${this.location})`);
+		}
+
+		this.textBlocks = questionBranch.text;
+		this.questions = questionBranch.questions;
+		this.questionLabels = {};
+
+		for (let i = 0; i < this.questions.length; i++) {
+			// Render labels
+			this.questions[i].label = await renderMarkdown(this.questions[i].label, undefined, true);
+			this.questionLabels[this.questions[i].name] = this.questions[i].label;
+			// Render options (if they exist)
+			let type = this.questions[i].type;
+			if (type === "checkbox" || type === "radio" || type === "select") {
+				for (let k = 0; k < this.questions[i].options.length; k++) {
+					this.questions[i].options[k] = await renderMarkdown(this.questions[i].options[k], undefined, true);
+				}
+			}
+		}
+
+		await this.loadSettings();
+		return this;
+	}
+
+	public async convertTo<T extends QuestionBranch>(type: keyof QuestionBranchTypes): Promise<T> {
+		// TODO typecast and return if types match - do not need to save to DB
+		await this.save();
+		await QuestionBranchConfig.update({ "name": this.name }, { "$set": { "type": type } });
+		// tslint:disable-next-line:no-use-before-declare
+		return await BranchConfig.loadBranchFromDB(this.name) as T;
+	}
+
+	public async save(): Promise<this> {
+		let branchConfig = await QuestionBranchConfig.findOne({ "name": this.name });
+		if (!branchConfig) {
+			// Insert into DB
+			await new QuestionBranchConfig({
+				"name": this.name,
+				"type": this.type,
+				"location": this.location,
```

Comment:
Just save the whole schema here. Only update the schema if it changes, this lets us update the real schema while keeping old questions around.